src/lib/stepper/stepper.ts

```diff
@@ -66,12 +69,23 @@ export class MatStep extends _MatStep implements ErrorStateMatcher {
 @Directive({
   selector: '[matStepper]'
 })
-export class MatStepper extends _MatStepper {
+export class MatStepper extends _MatStepper implements AfterContentInit, OnDestroy {
   /** The list of step headers of the steps in the stepper. */
   @ViewChildren(MatStepHeader, {read: ElementRef}) _stepHeader: QueryList<ElementRef>;
 
   /** Steps that the stepper holds. */
   @ContentChildren(MatStep) _steps: QueryList<MatStep>;
+  
+  /** Workaround for https://github.com/angular/material2/issues/8397 */
+  _stepChangesSubscription: Subscription;
+  ngAfterContentInit() {
+    this._stepChangesSubscription = this._steps.changes.subscribe(() => this._stateChanged());
+  }
+  ngOnDestroy() {
+    if(this._stepChangesSubscription instanceof Subscription) {
```

Comment:
A pattern we've started using for situations like this is to create a stream that emits on destory and then use `takeUntil` when describing (e.g. https://github.com/angular/material2/blob/master/src/lib/sidenav/drawer.ts#L467)

Reply:
So MatStep should emit on destroy and MatStepper should subscribe on each step?
To know which steps exist stepper would have to listen on QueryList changes...
Did I get that correctly?

Reply:
Like this, your `ngOnDestroy` would emit on a subject:
```ts
ngOnDestroy() {
  this._destroyed.next();
  this._destroyed.complete();
}
```

And then you don't have to save the `Subscription` when you subscribe, you can just use `takeUntil`:
```ts
this._steps.changes.pipe(takeUntil(this._destroyed)).subscribe(() => ...);
```

This way if we make more subscriptions later on we don't have to save them all as separate member variables